xmake/rules/vala/xmake.lua

```diff
@@ -58,31 +56,40 @@ rule("vala.build")
             target:add("sysincludedirs", path.directory(headerfile), {public = true})
         end
     end)
-    before_buildcmd_file(function (target, batchcmds, sourcefile_vala, opt)
+    before_buildcmd_files(function (target, batchcmds, sourcefiles_vala, opt)
+        -- Here we compile vala files into C code
+
+        -- We have to compile entire project each time
+        -- because otherwise valac can't resolve symbols
+        -- from other files, however, c files can be
+        -- incrementally built
 
         -- get valac
         import("lib.detect.find_tool")
         local valac = assert(find_tool("valac"), "valac not found!")
 
-        -- get c source file for vala
-        local sourcefile_c = target:autogenfile((sourcefile_vala:gsub(".vala$", ".c")))
-        local basedir = path.directory(sourcefile_c)
+        local argv = {"-C", "-d", target:autogendir()}
```

Comment:
`-d`? why not `-b basedir`?

Reply:
- `-b` outputs all files into single directory
- `-d` outputs all files according to source tree
- 
We need `-d` because of these two lines

```lua
local sourcefile_c = target:autogenfile((sourcebatch:gsub(".vala$", ".c")))
local basedir = path.directory(sourcefile_c)
```

In this case for files in sub-directories `basedir` is going to be a sub-directory, for which `-b` flag doesn't account. It worked before because it was assumed that files were compiled one-by-one and directory for `-b` flag was a `basedir`